### PR TITLE
add a formal structure for this document, similar to RFCs

### PR DESCRIPTION
**Pull request by staltz**
I've been browsing RFCs such as [webfinger](https://www.rfc-editor.org/rfc/rfc7033) and I noticed they typically have:

- Abstract
- Introduction
- Specification
- Examples (sometimes, not always, and it's separate from the Specification section)
- Security Considerations
- Normative References (other links you must abide by, "dependencies")
- Informative References (just other interesting and related links)

Also, this PR merged https://github.com/ssbc/ssb-meta-feeds-group-spec/pull/14 so we now have the Security Considerations section.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,86 +39,57 @@ two members. Thus we need to enable _sympathetic replication_, such that peers
...
-This work builds on the [ssb-meta-feeds-spec] (v1).
+Implementations supporting this specification MUST follow the
+[ssb-meta-feeds-spec] version 1.0.
 
 We define two types of feeds that each peer will have:
 
 1. A `group/additions` feed
 2. A "group feed" for each group
 
-```mermaid
```
Comment (staltz) [inline README.md:61]: The diagram was moved to the Examples section.

File: README.md [hunk 2]
```diff
@@ -347,9 +378,33 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+## 6. References
+
+### 6.1. Normative References
+
+- [ssb-meta-feeds-spec]
+- [envelope-spec]
+- [private-group-spec]
+- [ssb-meta-feeds-dm-spec]
+
+### 6.2. Informative References
```
Comment (arj03) [inline README.md:454]: here it might be good to link to other (non-ssb) group systems. Such as MLS (https://messaginglayersecurity.rocks/). There should be some notes in our chat about them in decentralized settings. I can try and find them again and add here. Not that it should block this PR. The structure is great and we can always add more information.

Comment (staltz) [reply README.md:454]: Agree, although I would want to put MLS reference (and a comparison discussion) in the "private groups spec" while this spec is just about the metafeeds+groups. I actually really like [original_notes.md](https://github.com/ssbc/envelope-spec/blob/master/original_notes.md) as a document and prefer it over the actual envelope-spec and private-group-spec, because it tells a story instead of being just bullet lists.

Comment (arj03) [reply README.md:454]: Ah right. Make sense. Good we have the original notes :)

File: README.md [hunk 3]
```diff
@@ -27,153 +39,113 @@ two members. Thus we need to enable _sympathetic replication_, such that peers
...
-  ],
-  tangles: {
-    group: {
-      // ...
-    },
-    members: {
-      // ...
-    }
-  }
-}
```
Comment (mixmix) [inline README.md:134]: I would personally like this example message to stay. You can hide it in a details fold if you want.
I don't want us to correlate "dry and impenetrable" with "looks academic".

Note that your spec dropped the `version` field.

What this example captures is that it really bangs hold which sort of ssb-uri should b e used, and what's meant by "base64 encoded string". An alternative to this would be to store such examples as test-vectors and link to those from the text.

File: README.md (hunk 1 above)
Comment (mixmix) [reply README.md:61]: I'm not in favour of moving the diagram. For me when reading a spec it's all about having a quick frame to hang the details on. If you info-dump text details on me it's very hard to parse and make sense.

Alt solution - link down in document "see example meta-feed structure"

_I will not die on this hill, I just want this to be accessible for other people like me_

File: README.md (hunk 3 above)
Comment (mixmix) [reply README.md:134]: **Summary**:
- :fire: MUST add version
    - we're jumping to v2 anyway - after we resolve this https://github.com/ssbc/private-group-spec/pull/23
    - could change `groupKey -> secret` while we're there
- :hot_pepper: SHOULD add tangles (see other spec?)
- <preference> would like example preserved in some way

File: README.md [hunk 4]
```diff
@@ -347,9 +378,77 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+symmetric encryption and static keys (no key rotation), which means it cannot
+guarantee [perfect-forward-secrecy] neither [post-compromise-security]. This
+makes SSB Private Groups less likely to guarantee confidentiality at scale, when
+a group has dozens or hundreds of members. If any of the members leaks the group
+secret, all past communication in the group is compromised.
+
+This also means that any group member can add new members to the group. They may
+even do so without publishing a `group/add-member` message, which effectively
+allows any group member to invite a third party that will remain undetected by
+existing group members.
```
Comment (mixmix) [inline README.md]:
```suggestion
existing group members. The only signal you might get is that a peer not in the group is asking for
group leaf feeds that they should not know about.
```

File: README.md [hunk 5]
```diff
@@ -347,9 +378,77 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+
+SSB Private Groups with several members should thus not be considered
+confidential nor forward secret, and should be used with caution. Our security
+model assumes that the group is as trustworthy as the _least_ trustworthy member
+in the group.
+
+The choice of symmetric encryption (as opposed to, e.g. hash-ratchet or
+double-ratchet) was made because of efficiency. Symmetric encryption means
+adding new members or publishing new content is `O(1)` fast, and the system is
+overall simple to implement.
```
Comment (mixmix) [inline README.md:406]: Actually, the decision was made because:
- double ratchet is harder to implement (we designed to be able to support it later ...eg. box3?)
- in Ahau's use case we NEED to be able to read all messages in a group, so fwd secrecy is not a desired feature

File: README.md [hunk 6]
```diff
@@ -347,9 +378,77 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+a group has dozens or hundreds of members. If any of the members leaks the group
+secret, all past communication in the group is compromised.
+
+This also means that any group member can add new members to the group. They may
+even do so without publishing a `group/add-member` message, which effectively
+allows any group member to invite a third party that will remain undetected by
+existing group members.
+
+SSB Private Groups with several members should thus not be considered
+confidential nor forward secret, and should be used with caution. Our security
```
Comment (mixmix) [inline README.md]: I think "not confidential" is a bit rough.

> Confidential - Done or communicated in confidence; secret.

Can we guarentee secrecy? No. No encryption system guarentees that to my knowledge - there is always a way to leak. You might argue can you leak but be plausably deniable? here no, because of the signatures of messages, don't know about other things like Signal? Signal is FOSS, could I make a hacked version which leaked? probably? You'd just have to never forget the ratchet keys?

File: README.md [hunk 7]
```diff
@@ -347,9 +378,77 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+
+### 5.2. Membership analysis attacks
+
+A group feed is discoverable only by the group members, because its announcement
+is encrypted. However, an eavesdropper who can replicate a peer's metafeed tree
+will notice the presence of these encrypted message on shard feeds. If the
+eavesdropper can do that for several peers, then they may perform a metadata
+analysis attack where they try to correlate the encrypted messages on the shard
+feeds to the group members.
+
```
Comment (mixmix) [inline README.md:416]:
```suggestion

In the current implementation, we mitigated through how shards are assigned
per user - because the derivation involves a hash function and a piece of
content unique to the peer (their root feed id), each peer will have the group
located in a different shard. This makes it harder to guess which group an 
announce message was for. 
```

Comment (mixmix) [reply README.md:416]: this will need weaving in a little more, but I think we should mention this, it's a security feature

We can also say timing attacks could be possible, but because there are unknown delays in the network, and in when people "accept" and set up a group feed, it makes it harder.
We could also further mitigate by only setting up the group feed when you go to publish in the group for the first time - adds even more delay.

File: README.md [hunk 8]
```diff
@@ -347,9 +378,77 @@ knows Staltz is part of the group because all `group/add-member` messages have
...
+This can be mitigated by publishing dummy encrypted messages on the shard feed
+at random intervals, with the downside of increasing the size of the shard feed,
+and thus making partial replication heavier.
+
+It is currently unknown how much can be learned from such analysis attack, as
+it depends on the network's complexity, such as number of peers, number of
+groups, and number of relationships between peers.
+
```
Comment (mixmix) [inline README.md:443]:
```suggestion

Another way membership information can leak is through replication requests -
if only Alice, Bob, and Carol ask for feeds A, B, C, and these feeds don't seem to
be associated with any publicly declared sub-feeds, it might be reasonable to guess
that Alice, Bob, Carol are all in a group and that A, B, C are their associated feeds
for that group. We think this can be mitigated through "sympathetic replication"
(asking friends who aren't in our groups to replicate some of our group feeds) in order
to add more noise (and replication resilience) to the system. This needs further
investigation and specification.
```

**Review by mixmix — CHANGES_REQUESTED** [review]
:fire: the `group/add-member` version needs pinning down.

The rest is looking good to me. I made some further suggestions to extend some areas.
Nice work :bowling:

File: README.md (hunk 4 above)
Comment (staltz) [reply README.md]: Good point

File: README.md (hunk 5 above)
Comment (staltz) [reply README.md:406]: > double ratchet is harder to implement

Doesn't that mean the same as

> the ("our") system is overall simple to implement

?

File: README.md (hunk 1 above)
Comment (staltz) [reply README.md:61]: Well, we got feedback that the spec was too informal and example driven, so ... we have to do something to make it less informal and example driven. Looking at RFCs, I don't see them writing with examples first, I see that examples are often Appendix or something like that.

I agree the diagram is nice :smiling_face_with_tear: But this is a spec, not a tutorial.

File: README.md (hunk 3 above)
Comment (staltz) [reply README.md:134]: I don't understand why we *have* to add the version, I understand why we *should* add it but technically things are going to still work just fine if peers implement only what this spec is saying. Anyway, I'll just add it back, just because you're asking.

However, let's not block this PR on https://github.com/ssbc/private-group-spec/pull/23 because that one is not done yet, and we're trying to *remove* obstacles for milestone completion, not add obstacles.

Comment (staltz) [reply README.md:134]: > What this example captures is that it really bangs hold which sort of ssb-uri should b e used, and what's meant by "base64 encoded string". An alternative to this would be to store such examples as test-vectors and link to those from the text.

I think we can just delegate this part of the spec to `private-group-spec`, so we don't need to "spec" it in two places.

File: README.md (hunk 8 above)
Comment (staltz) [reply README.md:443]: Good point

File: README.md (hunk 6 above)
Comment (staltz) [reply README.md]: I removed the "not confidential" and simplified this paragraph.

File: README.md (hunk 3 above)
Comment (staltz) [reply README.md:134]: Well, whatever, I put version `v2` and `secret` instead of groupKey. Please work on https://github.com/ssbc/private-group-spec/pull/23 ASAP
